projects/mercury/src/llm/Conversation.js

```diff
@@ -0,0 +1,415 @@
+        new FulltextAPI().getConversation(id).then(conversation => processResponseArticles(conversation));
+    };
+
+    const restoreChat = id => {
+        new FulltextAPI()
+            .getHistory(id)
+            .then(processHistoryResponseMessages)
+            .then(() => setLoading(false))
+            .catch(() => handleHttpError('Error retrieving chat history.'));
+    };
+
+    const processConversationHistory = data => {
+        if (data && data.length > 0) {
+            data.sort((a, b) => new Date(b.start) - new Date(a.start));
+
+            if (JSON.stringify(data) !== JSON.stringify(conversationHistory)) {
+                setConversationHistory(data);
+            }
+        }
+    };
+
+    const getAllConversationHistory = () => {
+        new FulltextAPI()
+            .getAllConversations()
+            .then(processConversationHistory)
+            .catch(() => {
+                handleHttpError('Error retrieving chat history.');
+                setLoading(false);
+            });
+    };
+
+    const prepareRestoreChat = id => {
+        setLoading(true);
+        setQuery('');
+        setResponseArticles([]);
+        setMessages([]);
+        setConversationId(id);
+        setRestoreChatStatus(true);
+        getAllConversationHistory();
+    };
+
+    const startNewConversation = () => {
+        setQuery('');
+        setResponseMessage('');
+        setMessages([]);
+        setResponseArticles([]);
+        setConversationId('');
+        getAllConversationHistory();
+    };
+
+    const deleteChat = id => {
+        new FulltextAPI()
+            .deleteChat(id)
+            .then(getAllConversationHistory)
+            .then(startNewConversation)
+            .then(() => setLoading(false))
+            .catch(() => handleHttpError('Error deleting chat history with id ' + id));
+    };
+
+    const processSearchQueryChange = newQuery => {
+        if (newQuery !== query) {
+            if (responseMessage !== '') {
+                setResponseMessage('');
+                setResponseArticles([]);
+            }
+
+            setQuery(newQuery);
+        }
+    };
+
+    const handleSearchError = error => {
+        setResponseMessage(error);
+        setLoading(false);
+    };
+
+    const performSearch = () => {
+        if (conversationId === '') {
+            const searchResult = new FulltextAPI().search(query);
+            return searchResult;
+        }
+
+        return new FulltextAPI().chat(query, conversationId);
+    };
+
+    const prepareFetchSearch = () => {
+        setLoading(true);
+        setResponseMessage('');
+
+        if (query === '') {
+            return;
+        }
+
+        performSearch()
+            .then(processResponseMessages)
+            .then(() => setLoading(false))
+            .catch(() => {
+                handleHttpError('Connection error.');
+                handleSearchError('Fairspace article search is not available at the moment');
```

Comment:
"Fairspace article search" - is this always articles? If so, we should make it more specific, otherwise we need more generic naming